subsys/debug/tracing/cpu_stats.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_cpu_stats.h>
+
+enum cpu_state {
+	CPU_STATE_IDLE,
+	CPU_STATE_NON_IDLE,
+	CPU_STATE_SCHEDULER
+};
+
+static volatile enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
+static volatile enum cpu_state cpu_state_before_interrupts;
+
+static volatile u32_t last_time;
+static volatile u64_t idle_cnt;
+static volatile u64_t non_idle_cnt;
+static volatile u64_t sched_cnt;
+static volatile int nested_interrupts;
+static struct k_thread *current_thread;
+
+#ifndef CONFIG_SMP
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
+{
+#ifdef CONFIG_SMP
+	return thread->base.is_idle;
+#else
+	return thread == _idle_thread;
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
+{
+	int key = irq_lock();
+	u32_t time = k_cycle_get_32();
+
+	if (time >= last_time) {
+		(*cnt) += (time - last_time);
+	} else {
+		(*cnt) += (UINT32_MAX - last_time + 1 + time);
+	}
+	last_time = time;
+	irq_unlock(key);
+}
+
+static void cpu_stats_update_counters(void)
+{
+	switch (last_cpu_state) {
+	case CPU_STATE_IDLE:
+		increase_counter(&idle_cnt);
+	break;
+
+	case CPU_STATE_NON_IDLE:
+		increase_counter(&non_idle_cnt);
+	break;
+
+	case CPU_STATE_SCHEDULER:
+		increase_counter(&sched_cnt);
+	break;
+
+	default:
+		__ASSERT_NO_MSG(false);
+	break;
+	}
+}
+
+u32_t cpu_stats_sched_get(void)
+{
+	cpu_stats_update_counters();
+	/* Dividing to get result in ms */
+	return SYS_CLOCK_HW_CYCLES_TO_NS(sched_cnt) / 1000000;
+}
+
+u32_t cpu_stats_idle_get(void)
+{
+	cpu_stats_update_counters();
+	/* Dividing to get result in ms */
+	return SYS_CLOCK_HW_CYCLES_TO_NS(idle_cnt) / 1000000;
+}
+
+u32_t cpu_stats_non_idle_get(void)
+{
+	cpu_stats_update_counters();
+	/* Dividing to get result in ms */
+	return SYS_CLOCK_HW_CYCLES_TO_NS(non_idle_cnt) / 1000000;
+}
+
+u32_t cpu_stats_non_idle_percent_get(void)
+{
+	cpu_stats_update_counters();
+	return ((non_idle_cnt + sched_cnt) * 100) /
+		(idle_cnt + non_idle_cnt + sched_cnt);
+}
+
+void cpu_stats_reset_counters(void)
+{
+	int key = irq_lock();
+
+	idle_cnt = 0;
+	non_idle_cnt = 0;
+	sched_cnt = 0;
+	last_time = k_cycle_get_32();
+	irq_unlock(key);
+}
+
+void sys_trace_thread_switched_in(void)
+{
+	__ASSERT_NO_MSG(nested_interrupts == 0);
+
+	cpu_stats_update_counters();
+	current_thread = k_current_get();
+	if (is_idle_thread(current_thread)) {
+		last_cpu_state = CPU_STATE_IDLE;
+	} else {
+		last_cpu_state = CPU_STATE_NON_IDLE;
+	}
+}
+
+void sys_trace_thread_switched_out(void)
+{
+	__ASSERT_NO_MSG(nested_interrupts == 0);
+	__ASSERT_NO_MSG(current_thread == k_current_get());
+
+	cpu_stats_update_counters();
+	last_cpu_state = CPU_STATE_SCHEDULER;
+}
+
+void sys_trace_isr_enter(void)
+{
+	if (nested_interrupts == 0) {
```

Comment:
this is very tricky note that you can be preempted by another interrupt before you increment the `nested_interrupts` (unless interrupts were disabled but I don't think they are). You cannot simply use atomic increment or cmpxch as you may end up with counters updated after the high priority interrupts are done. I think the only option is to lock interrupts for entire operation. (Note that you already lock interrupts while updating the counters.)

Reply:
done